app/scripts/models/account.js

```diff
@@ -298,24 +312,37 @@ define([
           return fxaClient.changePassword(email, oldPassword, newPassword);
         })
         .then(function () {
-          // sign the user in, keeping the current sessionTokenContext. This
-          // prevents sync users from seeing the `sign out` button on the
-          // settings screen.
           return fxaClient.signIn(
             email,
             newPassword,
             relier,
             {
-              reason: fxaClient.SIGNIN_REASON.PASSWORD_CHANGE,
-              sessionTokenContext: self.get('sessionTokenContext')
+              reason: fxaClient.SIGNIN_REASON.PASSWORD_CHANGE
             }
           );
         })
         .then(function (updatedSessionData) {
-          self.set(updatedSessionData);
+          self.storeSessionData(updatedSessionData, relier);
         });
     },
 
+    storeSessionData: function (sessionData, relier) {
+      var self = this;
+
+      sessionData.isSessionSharedWithFirefox =
+        self.get('isSessionSharedWithFirefox') || relier.isSync();
```

Comment:
Perhaps we can just pass the result of `relier.isSync()` so this method doesn't have to know about `relier`.